### PR TITLE
Components: Remove global button styles from Post Comment Actions

### PR DESCRIPTION
**Pull request by sarayourfriend**
#### Changes proposed in this Pull Request

* Use the `Button` component for the comment action buttons

#### Testing instructions

* Go to the reader and open a post's comments section. check that the buttons work and look like they do in production.

Related to #45259

### CODE REVIEW

File: client/assets/stylesheets/shared/_extends.scss
Comment (tyxla) [inline client/assets/stylesheets/shared/_extends.scss:51]: Let's add a newline at the end of the file.

**Review by tyxla — CHANGES_REQUESTED** [review]
There seems to be a subtle difference between the original spacing above/below the comment action buttons, and the one in this branch.

Before:


After:


As you can see, with the spacing in this PR, comment action buttons aren't vertically centered, while originally they are more or less vertically centered.

**Comment by jsnajdr** [comment]
> There seems to be a subtle difference between the original spacing above/below the comment action buttons

This patch adds the `.button` class to the comment action buttons, and there are two CSS properties that modify the spacing.

After the patch, the space above the button row is smaller, i.e., the button row is closer to the text above it. That's because of:
```css
.button .gridicon {
  margin-top: -2px;
}
```
which adds negative margin on top of the gridicon element, making the overall `.button` box height smaller. As the box text baseline is aligned with the bottom edge of the box, it removes whitespace above the text.

After the patch, the space below the button row is bigger, i.e., there's more space above the next comment. That's because of:
```css
.button {
  line-height: 22px;
}
```
The button has a bigger `line-height` that it used to. This, in theory, adds more space both above and below the baseline. In this case, it adds space only below the baseline, probably because the part of the box above the baseline is already tall enough because of the gridicon element.

If we want to rebalance the spacing again, the best solution seems to be adding a `margin-top` to the `.comments__comment-actions` container.

File: client/blocks/comments/comment-actions.scss
Comment (jsnajdr) [inline client/blocks/comments/comment-actions.scss:9]: Is this extend really needed? I would expect the element get all the needed styles from the `.button` class.

Looking at devtools, I see some differences in, e.g., `font-size` and `padding`. But I'm a bit confused by all that rather than understanding the intent:

Screenshot 2020-09-02 at 14 32 19

Comment (saramarcondes) [reply client/blocks/comments/comment-actions.scss:9]: > Looking at devtools, I see some differences in, e.g., font-size and padding. But I'm a bit confused by all that rather than understanding the intent:

The problem is that you need to also remove the global button styles for this to make sense. If you remove the global styles, there are several issues that arise. Therefore, I added the extends in order to be able to maintain those styles while removing one thing from depending on them globally. Future change sets can continue to extend `%button` until we've fully removed the dependency on the global application of those styles to all buttons.

Right now, you're seeing both the global and the local application of those styles, so it's not going to make a lot of sense (it'll look like duplication mostly). If you remove the global button styles from `_reset.scss` it might make more sense.

On a separate note, I think I should actually open a PR to fix this in the component itself, because it currently relies on the Calypso resets as well and should have those styles built in. If you remove the resets, even the `Button` component will look bad:

Captura de Tela 2020-09-02 às 09 49 47

Comment (saramarcondes) [reply client/blocks/comments/comment-actions.scss]: Thank you for pointing this out. I was indeed going about this the wrong way. I've updated the branch with new code that doesn't rely on the global styles.

The problem I was running into was that I was removing the global styles to check how my changes were working, and getting confused with the like button, which was using a different method of getting the `button` tag. I've added something of a hack to make it work using the `Button` component, but I'm not sure if there's a better way to do this. I've never worked with this `tagName` pattern in React before, so I'm kind of just guessing based on what I can imagine will work based on the implementation I am seeing. I would especially love feedback with respect to that.

**Comment by saramarcondes** [comment]
@tyxla I believe there are still _slight_ differences, but I think the new version is actually more consistent with respect to how the icons are lined up. Let me know if you think there's still an issue though and I can try to match it more exactly.

**Review by tyxla — COMMENTED** [review]
Thanks for your continuous work on this one 👍 I wouldn't want to block this PR any further, but I've noticed 2 more things I'd like to talk through before approving it:

* Spacing-wise, it's not ideal, but I think it's indeed closer to what we want. One thing I'm not super happy about is the spacing on the left - it doesn't align with the comment text above, so we might want to remove that left padding on the read more button. What are your thoughts?
* Color-wise, there's a difference between production and this branch for the read more button.

Production:


This PR:


Is that change intentional?

**Review by jsnajdr — COMMENTED** [review]
@saramarcondes I think we're dealing with three kinds of buttons here:

1. `button` with native browser styles that make it look like, well, a button 🙂
2. Calypso `<Button>` that adds additional styling on top of the browser native ones (without any reset) and makes the button still look like button, but a Calypso-styled one (borders, colors, paddings, font, focus rings...)
3. `button` element that is styled to look like a regular `div`, i.e., one that resets the native browser styles. That's useful at places where we want `button` behavior (with all the a11y benefits it provider over a `<div onClick>`), but none of the styles. The comments actions buttons are good example of this: they're styled to look more like text links.

The `%button` extend you had in the previous version of the patch provided these button-to-div resets.

`<Button borderless>` is close to the button-as-div look, but not quite there. It's intended to still have the same dimensions, paddings and fonts like the regular button, just without the border. See how the borderless and border-ful buttons nicely align in this toolbar screenshot:

Screenshot 2020-09-03 at 11 36 23

A button-as-div wouldn't do that. The `.comments__comment-actions button.button` CSS rule needs to do several somewhat random overrides to get from `borderless` to the desired button-as-div state.

Ideally, I think we need something like `<Button plain>` or `<Button bare>` to get that button-as-div styling with the reset. Such a button wouldn't have the `.button` CSS class, as that's already used for the Calypso-styled button. Instead, there could be a `.button-plain` CSS class that does the reset -- it would have the same contents as `%button` had.

I don't want to block this PR, once it gets good enough to ship, until we implement some perfect solution, but let's discuss if what I write above makes sense and if it's a viable roadmap to make the Calypso buttons a bit better.

@diegohaz do you have any insights about how to do the button-as-div components and how, e.g., Reakit or `@wordpress/components` approach this?

File: client/blocks/comments/comment-actions.scss
Comment (jsnajdr) [inline client/blocks/comments/comment-actions.scss:10]: This style conflicts with the one that gives the "Read more" button the primary color:
```css
.comments__comment-actions .comments__comment-actions-read-more {
	color: var( --color-primary );
}
```

That causes the coloring issue that @tyxla reports in https://github.com/Automattic/wp-calypso/pull/45344#pullrequestreview-481633015

**Comment by saramarcondes** [comment]
@jsnajdr Thanks for the detailed thoughts. I agree with you that we should take a different approach for this. I've added the `bare` style in this PR as you suggested and will revisit this PR once that is merged: https://github.com/Automattic/wp-calypso/pull/45386

**Review by tyxla — APPROVED** [review]
Looks and works well to me, and the code looks good.

On a side note, I'm not sure `bare` makes complete sense in terms of being predictive, obvious and clear what it does. What alternatives of `bare` can we find? Material design calls those "text buttons". while Bootstrap calls them Links or Link Buttons (it also has a Jumbotron component, so not sure we wanna borrow naming ideas from there, lol). Chakra UI also calls them links FWIW.

What are your thoughts on the prop naming?

**Comment by tyxla** [comment]
FWIW, consider the `bare` discussion irrelevant to this PR's blocked state. From my perspective, this is good to go already.

**Comment by jsnajdr** [comment]
> What are your thoughts on the prop naming?

I'd rule out "link buttons" because that suggests the component is an `<a href>` link with button styling. While it's exactly the opposite: it's a `<button>` without button styling 🙂 `<Button text>` or `<Button plain>` look like good alternatives to me.

**Comment by tyxla** [comment]
I agree on all points @jsnajdr. While `text` is a bit ambiguous, `plain` sounds great to me FWIW 👍

**Comment by diegohaz** [comment]
> @diegohaz do you have any insights about how to do the button-as-div components and how, e.g., Reakit or `@wordpress/components` approach this?

Reakit doesn't care about styling. The closest concept is the `as` prop that can be used to change the underlying element. It's not about styling, but how the element will behave. You can do something like `<Button as="div">`, which would render a `<div>` element with all the necessary behaviors so it remains accessible:

- `role="button"` and `tabindex` attributes are added to the div.
- Pressing <kbd>Enter</kbd> triggers click on the div on key down.
- Pressing <kbd>Space</kbd> activates the button (similar to `:active`) on key down and triggers click on key up.
- `disabled` adds `aria-disabled` and disables pointer and keyboard events on the div.

In `@wordpress/components`, a `Button` is by default a pretty simple element with most browser styles reset, including `appearance`: https://wordpress.github.io/gutenberg/?path=/story/components-button--default

It has props like [`isPrimary`](https://wordpress.github.io/gutenberg/?path=/story/components-button--primary), [`isSecondary`](https://wordpress.github.io/gutenberg/?path=/story/components-button--secondary), [`isTertiary`](https://wordpress.github.io/gutenberg/?path=/story/components-button--tertiary). And they are responsible for making it look more like a WordPress button.

The polymorphism there is resumed to the `button` and `a` elements.

**Comment by saramarcondes** [comment]
@tyxla I'd love a re-review after having replaced the Automattic component button with the WordPress components equivalent. Let me know what you think.

**Review by tyxla — COMMENTED** [review]
Thanks for using `@wordpress/components` instead, @saramarcondes!

FWIW, currently in calypso, we don't add the `@wordpress/components` stylesheet, so any WordPress components you use won't have their corresponding styles loaded. 

#45724 is taking care of that, so maybe we should rebase and test this one again after we land #45724. What do you think?

**Comment by saramarcondes** [comment]
@tyxla Oh gosh, I feel very silly! Yes, that would make sense to do first, wouldn't it 🤦‍♀️ It didn't even occur to me that we needed to load the stylesheet separately or that it would be affecting these PRs. Not sure how I didn't connect those dots.

Lets land #45724 like you said and then revisit this one.

**Comment by tyxla** [comment]
@saramarcondes I've rebased the PR since we merged https://github.com/Automattic/wp-calypso/pull/45724 and it seems like the buttons work really well! 

The only thing we needed to do is to reset the `height` that's coming by default for `@wordpress/components` buttons. I've taken care of this in https://github.com/Automattic/wp-calypso/pull/45344/commits/982ea2d88f27920f5160a02a2695735bb29258f1 and this PR is good to go from my perspective. 

It's worth noting that I've also tested the moderation actions, which are currently disabled in the UI. Regardless, they work just fine with this PR.

**Review by tyxla — APPROVED** [review]
LGTM! Nice work 👏

**Comment by tyxla** [comment]
Going to help ship this once tests are green.

**Pull request merged**
